File: heaps/ipo.py
import heapq


def maximize_capital(k: int, wealth: int, profits: list[int], capitals: list[int]) -> int:  # LeetCode 502.
    """Select at most k distinct projects to maximize wealth."""
    capitals_profits: list[tuple[int, int]] = [
        (iter_capital, profit) for iter_capital, profit in zip(capitals, profits)
    ]
    capitals_profits.sort()

    profits_heap = []
    while k > 0 and (capitals_profits or profits_heap):
        while capitals_profits and wealth >= capitals_profits[0][0]:
            # Negate profit to fit into max heap.
            heapq.heappush(profits_heap, -capitals_profits.pop(0)[1])

        if not profits_heap:  # All remaining projects are unaffordable.
            break

        wealth -= heapq.heappop(profits_heap)  # Negate profit back.
        k -= 1

    return wealth

File: heaps/ipo/ipo.cpp
#include <vector>
#include <queue>
using namespace std;

int maximize_capital(int k, int wealth, vector<int> &profits, vector<int> &capital)
{ // LeetCode 502.
    // Min heap. Format: {capital, profit idx}.
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> capitals_heap;
    for (int idx = 0; idx < capital.size(); idx++)
    {
        capitals_heap.push({capital[idx], idx});
    }

    priority_queue<int> profits_heap; // Max heap.
    while (k > 0)
    {
        while (!capitals_heap.empty() && wealth >= capitals_heap.top().first)
        {
            // Grab all the affordable projects' profits into max heap.
            auto [capital, profit_idx] = capitals_heap.top();
            profits_heap.push(profits[profit_idx]);
            capitals_heap.pop();
        }

        if (profits_heap.empty())
        { // No more affordable projects.
            break;
        }
        else
        {
            wealth += profits_heap.top();
            k -= 1;
            profits_heap.pop();
        }
    }

    return wealth;
}
